File: Algorithms/0542_01_Matrix/Python/01_Matrix_Solution_1.py

# Space: O(1)
# Time: O(n)

class Solution:
    def updateMatrix(self, mat):
        m = len(mat)
        n = len(mat[0])
        if m == 1 and n == 1: return mat

        for y in range(m):
            for x in range(n):
                if mat[y][x] != 0:
                    left = mat[y][x - 1] if x - 1 >= 0 else float('inf')
                    top = mat[y - 1][x] if y - 1 >= 0 else float('inf')
                    mat[y][x] = min(top, left) + 1

        for y in range(m - 1, -1, -1):
            for x in range(n - 1, -1, -1):
                if mat[y][x] != 0:
                    right = mat[y][x + 1] if x + 1 < n else float('inf')
                    bottom = mat[y + 1][x] if y + 1 < m else float('inf')
                    mat[y][x] = min(right + 1, bottom + 1, mat[y][x])
        return mat

File: Algorithms/0542_01_Matrix/Java/01MatrixSolution1.java

// Space: O(1)
// Time: O(n)

class Solution {
    public int[][] updateMatrix(int[][] mat) {
        int m = mat.length;
        int n = mat[0].length;
        int maxValue = 1000000;
        if (m == 1 && n == 1) return mat;
        for (int y = 0; y < m; y++) {
            for (int x = 0; x < n; x++) {
                if (mat[y][x] != 0) {
                    int left = x - 1 >= 0 ? mat[y][x - 1] : maxValue;
                    int top = y - 1 >= 0 ? mat[y - 1][x] : maxValue;
                    mat[y][x] = Math.min(left + 1, top + 1);
                }
            }
        }
        for (int y = m - 1; y >= 0; y--) {
            for (int x = n - 1; x >= 0; x--) {
                int right = x + 1 < n ? mat[y][x + 1] : maxValue;
                int bottom = y + 1 < m ? mat[y + 1][x] : maxValue;
                mat[y][x] = Math.min(mat[y][x],Math.min(right+1,bottom+1));
            }
        }
        return mat;
    }
}




